Implementation of Wompi Gateway.

fbrusatti opened the pull request:
Here is a first round to implement Wompi gateway.

There are implemented

  * `store` method to get a token to use in subsequent requests
  * `purchase` to run a transaction (they have output PENDING generally)
  * `query_transaction` to see how is the status of transactions
  * `pse_financial_institutions` to query the financial institutions availables

ref: https://docs.wompi.co/docs/en/inicio-rapido

Here the remote test
```
$ ruby -Itest test/remote/gateways/remote_wompi_test.rb 
Loaded suite test/remote/gateways/remote_wompi_test
Started
........
Finished in 24.592320023 seconds.
-----------------------------------------------------------------------------------------------------------------------
8 tests, 16 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
-----------------------------------------------------------------------------------------------------------------------
0.33 tests/s, 0.65 assertions/s
```

Ref waysact/evergiving#6379

fbrusatti committed:
Implementation of Wompi Gateway. Ref waysact/evergiving#6379

fbrusatti committed:
Implementation of Wompi Gateway. Ref waysact/evergiving#6379

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,246 @@
+
+      def scrub(transcript)
+        transcript.
+          gsub(/(Authorization: Bearer )([A-Za-z0-9\-\._~\+\/]+=*)/, '\1[FILTERED]').
```
This should filter the Bearer token regardless of its format.

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,246 @@
+
+      def add_customer_data(post, options)
+        post[:customer_data] = {}
+        post[:customer_data][:phone_number] = options[:customer]&.[](:mob_phone)
```
This could be more readable with using `dig`

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,246 @@
+        {
+          'accept' => '*/*',
+          'Content-Type' => 'application/json',
+          'Authorization' => "Bearer #{@options.send(:[], key)}"
```
Should this fail fast if there's no auth token?

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,246 @@
+      end
+
+      def reference_in_params?(action)
+        action.match(/\A?reference=[^&]+&*/).present?
```
`\A?` is redundant, or did you mean to escape the `?`?

Might be more robust to use `CGI::parse` here.

jscheid reviewed (commented):
(Removed)

fbrusatti committed:
Improve source code to do it resilient.

  + scrub Bearer token without care of token format
  + Use dig instead safety navigation operator (&.) to improve
    readability
  + Fail early if there is no Bearer token present
  + Use CGI::parse to do it more resilient

Ref waysct/evergiving#6379

fbrusatti committed:
Improve source code to do it resilient.

  + scrub Bearer token without care of token format
  + Use dig instead safety navigation operator (&.) to improve
    readability
  + Fail early if there is no Bearer token present
  + Use CGI::parse to do it more resilient

Ref waysct/evergiving#6379

fbrusatti committed:
Expose query acceptance token as a new gateway endpoint. Ref waysact/evergiving#6379

fbrusatti committed:
Expose query acceptance token as a new gateway endpoint. Ref waysact/evergiving#6379

fbrusatti committed:
Fix the way to return the acceptance token. Ref waysact/evergiving#6379

fbrusatti committed:
Fix the way to return the acceptance token. Ref waysact/evergiving#6379

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,249 @@
+
+      def scrub(transcript)
+        transcript.
+          gsub(/(Authorization: Bearer )([^\s])+/, '\1[FILTERED]').
```
`\S` matches non-whitespace:

```suggestion
          gsub(/\b(Authorization:\s+Bearer\s+)\S+/, '\1[FILTERED]').
```

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,249 @@
+        key = reference_in_params?(action) ? :private_key : :public_key
+        token = @options.send(:[], key)
+
+        raise 'Missing Bearer token' if token.nil?
```
```suggestion
        token = @options.fetch(key)
```

Is shorter and the error message says exactly which key was expected. However I just noticed that you check for options above with `requires!` so my original comment was wrong, you don't necessarily need to check for presence here. Then again it can't hurt to make the expectation clear here.

jscheid commented on lib/active_merchant/billing/gateways/wompi.rb:
```diff
@@ -0,0 +1,249 @@
+      end
+
+      def reference_in_params?(action)
+        path_params_ref?(action) || query_string_ref?(action.split('?').last)
```
```suggestion
        action_uri = URI(action)
        path_params_ref?(action_uri.path) || query_string_ref?(action_uri.query)
```

fbrusatti committed:
Update lib/active_merchant/billing/gateways/wompi.rb

Co-authored-by: Julian Scheid <julians37@gmail.com>

fbrusatti committed:
Update lib/active_merchant/billing/gateways/wompi.rb

Co-authored-by: Julian Scheid <julians37@gmail.com>

fbrusatti committed:
Merge branch '6379-wompi-gateway' of github.com:waysact/active_merchant into 6379-wompi-gateway

* '6379-wompi-gateway' of github.com:waysact/active_merchant:
  Update lib/active_merchant/billing/gateways/wompi.rb
  Update lib/active_merchant/billing/gateways/wompi.rb
  Fix the way to return the acceptance token. Ref waysact/evergiving#6379
  Expose query acceptance token as a new gateway endpoint. Ref waysact/evergiving#6379
  Improve source code to do it resilient.
  Implementation of Wompi Gateway. Ref waysact/evergiving#6379

fbrusatti committed:
Adjust puchase method to generate a store token and capture. Ref waysact/evergiving#6379

fbrusatti committed:
Rewrite test unit for Wompi & improve the message_from method. Ref waysact/evergiving#6379

fbrusatti committed:
Adjust Wompi gateweay to deal correctly with PSE transactions. Ref waysact/evergiving#6379
